crates/subspace-proof-of-time/src/pot_aes.rs

```diff
@@ -43,58 +39,21 @@ pub(crate) fn verify_sequential(
 ) -> bool {
     let key = GenericArray::from(PotBytes::from(*key));
     let cipher = Aes128::new(&key);
-    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
-
-    for checkpoint in checkpoints {
-        // Encrypt in place checkpoint_iterations times
-        // and compare with the checkpoint.
-        for _ in 0..checkpoint_iterations {
-            cipher.encrypt_block(&mut cur_block);
-        }
 
-        let checkpoint_block = GenericArray::from(PotBytes::from(*checkpoint));
-        if cur_block != checkpoint_block {
-            return false;
-        }
+    let mut inputs = Vec::with_capacity(checkpoints.len());
+    inputs.push(GenericArray::from(PotBytes::from(*seed)));
+    for checkpoint in checkpoints.iter().rev().skip(1).rev() {
+        inputs.push(GenericArray::from(PotBytes::from(*checkpoint)));
     }
-    true
-}
 
-/// Verifies the AES based proof in parallel.
-#[cfg(any(feature = "parallel", test))]
-pub(crate) fn verify_parallel(
-    seed: &PotSeed,
-    key: &PotKey,
-    checkpoints: &[PotCheckpoint],
-    checkpoint_iterations: u32,
-) -> bool {
-    let key = GenericArray::from(PotBytes::from(*key));
-    let cipher = Aes128::new(&key);
-
-    // Create the cipher pairs to be evaluated
-    let mut pairs = Vec::new();
-    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
-    for checkpoint in checkpoints {
-        let checkpoint_block = GenericArray::from(PotBytes::from(*checkpoint));
-        pairs.push((cur_block, checkpoint_block));
-        cur_block = checkpoint_block;
+    for _ in 0..checkpoint_iterations {
+        cipher.encrypt_blocks(&mut inputs);
```

Comment:
Would there be any difference in performance if this was using `decrypt_blocks(checkpoints)` and comparing to `inputs`?

Reply:
I was thinking about that, but I don't believe there is any performance difference between encoding and decoding instructions. But I do have plans to use decoding AND encoding at the same time to cut number of checkpoints in half (will do encoding and decoding in parallel and meet in the middle comparing results :exploding_head:).

Reply:
Interesting approach, makes sense. I thought there was some magic in `aesdec` instructions. Agree that 16 checkpoints is probably too much, 8 or even 4 if possible would be great.

Reply:
From my testing 16 checkpoints results in the fastest single-core verification. So we probably want 16, but by having 8 and doing half-encoding and half-decoding we can effectively get 16 (leveraging the fact that we have both inputs and outputs). The goal is to basically decrease proof size without decreasing verification performance.